msal/src/main/java/com/microsoft/identity/client/NativeAuthPublicClientApplication.kt

```diff
@@ -0,0 +1,115 @@
+import com.microsoft.identity.client.exception.MsalException
+import com.microsoft.identity.client.internal.CommandParametersAdapter
+import com.microsoft.identity.client.statemachine.BrowserRequiredError
+import com.microsoft.identity.client.statemachine.GeneralError
+import com.microsoft.identity.client.statemachine.InvalidAttributesError
+import com.microsoft.identity.client.statemachine.InvalidEmailError
+import com.microsoft.identity.client.statemachine.InvalidPasswordError
+import com.microsoft.identity.client.statemachine.PasswordIncorrectError
+import com.microsoft.identity.client.statemachine.UserAlreadyExistsError
+import com.microsoft.identity.client.statemachine.UserNotFoundError
+import com.microsoft.identity.common.crypto.AndroidAuthSdkStorageEncryptionManager
+import com.microsoft.identity.common.internal.cache.SharedPreferencesFileManager
+import com.microsoft.identity.common.internal.controllers.LocalMSALController
+import com.microsoft.identity.common.internal.controllers.NativeAuthMsalController
+import com.microsoft.identity.common.internal.net.cache.HttpCache
+import com.microsoft.identity.common.java.authorities.Authority
+import com.microsoft.identity.common.java.cache.ICacheRecord
+import com.microsoft.identity.common.java.commands.CommandCallback
+import com.microsoft.identity.common.java.controllers.CommandDispatcher
+import com.microsoft.identity.common.java.controllers.results.ICommandResult
+import com.microsoft.identity.common.java.eststelemetry.PublicApiId
+import com.microsoft.identity.common.java.exception.BaseException
+import com.microsoft.identity.common.java.logging.LogSession
+import com.microsoft.identity.common.java.logging.Logger.LogLevel
+import com.microsoft.identity.common.java.providers.microsoft.azureactivedirectory.AzureActiveDirectory
+import com.microsoft.identity.common.java.util.ResultFuture
+import com.microsoft.identity.common.java.util.checkAndWrapCommandResultType
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+/**
+ * NativeAuthPublicClientApplication provides implementation for the top level interface
+ * [INativeAuthPublicClientApplication] used by third party developers.
+ */
+class NativeAuthPublicClientApplication(
+    private val nativeAuthConfig: NativeAuthPublicClientApplicationConfiguration
+) : INativeAuthPublicClientApplication, PublicClientApplication(nativeAuthConfig) {
+
+    private lateinit var sharedPreferencesFileManager: SharedPreferencesFileManager
+
+    init {
+        initializeApplication()
+        initializeSharedPreferenceFileManager(nativeAuthConfig.appContext)
+    }
+
+    companion object {
+        /**
+         * Name of the shared preference cache for storing NativeAuthPublicClientApplication data.
+         */
+        private const val NATIVE_AUTH_CREDENTIAL_SHARED_PREFERENCES =
```

Comment:
I guess you could've considered extending your class from SingleAccountPublicClientApplication and that would've allowed you to leverage a lot of the single account functionality that is already built-in into that class.

Reply:
No, because our native auth version of `getCurrentAccount()` (part of a next PR) would conflict with the existing `getCurrentAccount()` (and we can't reuse the existing one, as the logic in ours is different). There would have been limited to no reusability of code benefit if we had extended `SingleAccountPublicClientApplication` instead of what we do now.

Reply:
What would you like me to expand on? We need a method called `getCurrentAccount()` which already exists in `SingleAccountPublicClientApplication`. It's not ideal that we're not extending this class, but without changing the existing MSAL code and SDK interface I don't see how this can be solved.